google/cloud/aiplatform/models.py

```diff
@@ -215,10 +228,483 @@ def upload(
+                automatic_resources=automatic_resources,
+                model=model.resource_name,
+                display_name=deployed_model_display_name,
+            )
+
+        if traffic_split is None:
+            if bool(self._gca_resource.traffic_split) is False:
+                raise ValueError(
+                    """There are currently no deployed models so the traffic percentage
+                for this deployed model needs to be 100."""
+                )
+            traffic_split = self._allocate_traffic(
+                traffic_split=dict(self._gca_resource.traffic_split),
+                traffic_percentage=traffic_percentage,
+            )
+        # else:
+        #     TODO: check traffic split provided is valid
+
+        operation_future = self.endpoint_client.deploy_model(
+            endpoint=self.resource_name,
+            deployed_model=deployed_model,
+            traffic_split=traffic_split,
+            metadata=metadata,
+        )
+
+        return lro.LRO(operation_future)
+
+    def undeploy(
+        self,
+        deployed_model_id: str,
+        traffic_split: Optional[Dict] = None,
+        metadata: Optional[Sequence[Tuple[str, str]]] = (),
+    ) -> lro.LRO:
+        """Undeploys a deployed model.
+
+        Proportionally adjusts the traffic_split among the remaining deployed
+        models of the endpoint.
+
+        Args:
+            deployed_model_id (str):
+                Required. The ID of the DeployedModel to be undeployed from the
+                Endpoint.
+            traffic_split (Sequence[gca_endpoint.Endpoint.TrafficSplitEntry]`):
+                Optional. If this field is provided, then the Endpoint's traffic_split
+                will be overwritten with it. If last DeployedModel is being
+                undeployed from the Endpoint, the [Endpoint.traffic_split] will
+                always end up empty when this operation completes. A DeployedModel
+                will be successfully undeployed only if it doesn't have any traffic
+                assigned to it. If this field is not provided, the traffic of the
+                remaining deployed models will be scaled to fill 100 percent.
+            metadata (Sequence[Tuple[str, str]]):
+                Optional. Strings which should be sent along with the request as
+                metadata.
+        Returns:
+            lro (lro.LRO):
+                LRO of model undeployment.
+        """
+        if traffic_split is None:
+            traffic_split = self._unallocate_traffic(
+                traffic_split=dict(self._gca_resource.traffic_split),
+                deployed_model_id=deployed_model_id,
+            )
+
+        operation_future = self.endpoint_client.undeploy_model(
+            endpoint=self.resource_name,
+            deployed_model_id=deployed_model_id,
+            traffic_split=traffic_split,
+            metadata=metadata,
+        )
+
+        return lro.LRO(operation_future)
```

Comment:
Block and probably no need to return the LRO.

Reply:
why block? will we change this in the future?

Reply:
Acked and reasoned above.